### ISSUE TITLE
gpg keyserver receive failes with "No data"

### ISSUE THREAD

**Issue by IPlayZed** [Morganamilo/paru#1114]
### Affected Version
**paru v2.0.1 - libalpm v13.0.2**

### Description
When trying to install any AUR package needing to import (a) GPG key (ABAF11C65A2970B130ABE3C479BE3E4300411886) paru fails to receive it. (I do not know if other keys are affected.)

~~Using manual package installation or something other (like yay for instance) works fine.~~
This fails any way, but specifying the gpg server when using gpg directly seems to work:
```
gpg --keyserver hkps://keys.openpgp.org --recv-keys ABAF11C65A2970B130ABE3C479BE3E4300411886
gpg: key 79BE3E4300411886: no user ID
gpg: Total number processed: 1
```

I tried running `paru -Sc` to clean cache, deleting all from `~/.cache/paru/clones`, running `paru -Sy archlinux-keyring && paru -Su`, adding `keyserver hkps://keys.openpgp.org` to `/etc/pacman.d/gnupg/gpg.conf`, running `gpg --keyserver hkps://keys.openpgp.org --recv-keys ABAF11C65A2970B130ABE3C479BE3E4300411886`, but none of them helped.

```
gpg: key 79BE3E4300411886: no user ID
gpg: Total number processed: 1
```

### Output

```sh
[system@arch-test linux-mainline]$ paru -S linux-mainline
:: Resolving dependencies...
Aur (1)                                     Old Version  New Version       Make Only
aur/linux-mainline                                       6.7rc6-1          No

:: Proceed to review? [Y/n]: 

:: Downloading PKGBUILDs...
 PKGBUILDs up to date
 nothing new to review
:: keys need to be imported:
     ABAF11C65A2970B130ABE3C479BE3E4300411886 wanted by: linux-mainline-6.7rc6-1
:: import? [Y/n]: 
gpg: keyserver receive failed: No data
error: failed to run: gpg --recv-keys ABAF11C65A2970B130ABE3C479BE3E4300411886:
```

pacman.conf
```
#
# /etc/pacman.conf
#
# See the pacman.conf(5) manpage for option and repository directives

#
# GENERAL OPTIONS
#
[options]
# The following paths are commented out with their default values listed.
# If you wish to use different paths, uncomment and update the paths.
#RootDir     = /
#DBPath      = /var/lib/pacman/
#CacheDir    = /var/cache/pacman/pkg/
#LogFile     = /var/log/pacman.log
#GPGDir      = /etc/pacman.d/gnupg/
#HookDir     = /etc/pacman.d/hooks/
HoldPkg     = pacman glibc
#XferCommand = /usr/bin/curl -L -C - -f -o %o %u
#XferCommand = /usr/bin/wget --passive-ftp -c -O %o %u
#CleanMethod = KeepInstalled
Architecture = auto

# Pacman won't upgrade packages listed in IgnorePkg and members of IgnoreGroup
#IgnorePkg   =
#IgnoreGroup =
![image](https://github.com/Morganamilo/paru/assets/21132316/34cb7594-42c2-40da-b712-771396aa6c60)

#NoUpgrade   =
#NoExtract   =

# Misc options
#UseSyslog
#Color
#NoProgressBar
CheckSpace
#VerbosePkgLists
#ParallelDownloads = 5

# By default, pacman accepts packages signed by keys that its local keyring
# trusts (see pacman-key and its man page), as well as unsigned packages.
SigLevel    = Required DatabaseOptional
LocalFileSigLevel = Optional
#RemoteFileSigLevel = Required

# NOTE: You must run `pacman-key --init` before first using pacman; the local
# keyring can then be populated with the keys of all official Arch Linux
# packagers with `pacman-key --populate archlinux`.

#
# REPOSITORIES
#   - can be defined here or included from another file
#   - pacman will search repositories in the order defined here
#   - local/custom mirrors can be added here or in separate files
#   - repositories listed first will take precedence when packages
#     have identical names, regardless of version number
#   - URLs will have $repo replaced by the name of the current repo
#   - URLs will have $arch replaced by the name of the architecture
#
# Repository entries are of the format:
#       [repo-name]
#       Server = ServerName
#       Include = IncludePath
#
# The header [repo-name] is crucial - it must be present and
# uncommented to enable the repo.
#

# The testing repositories are disabled by default. To enable, uncomment the
# repo name header and Include lines. You can add preferred servers immediately
# after the header, and they will be used before the default mirrors.

#[testing]
#Include = /etc/pacman.d/mirrorlist

[core]
Include = /etc/pacman.d/mirrorlist

[extra]
Include = /etc/pacman.d/mirrorlist

#[community-testing]
#Include = /etc/pacman.d/mirrorlist

[community]
Include = /etc/pacman.d/mirrorlist

# If you want to run 32 bit applications on your x86_64 system,
# enable the multilib repositories as required here.

#[multilib-testing]
#Include = /etc/pacman.d/mirrorlist

[multilib]
Include = /etc/pacman.d/mirrorlist

# An example of a custom package repository.  See the pacman manpage for
# tips on creating your own repositories.
#[custom]
#SigLevel = Optional TrustAll
#Server = file:///home/custompkgs
```

paru.conf
```
#
# $PARU_CONF
# /etc/paru.conf
# ~/.config/paru/paru.conf
#
# See the paru.conf(5) manpage for options

#
# GENERAL OPTIONS
#
[options]
PgpFetch
Devel
Provides
DevelSuffixes = -git -cvs -svn -bzr -darcs -always -hg -fossil
#AurOnly
#BottomUp
#RemoveMake
#SudoLoop
#UseAsk
#SaveChanges
#CombinedUpgrade
#CleanAfter
#UpgradeMenu
#NewsOnUpgrade

#LocalRepo
#Chroot
#Sign
#SignDb
#KeepRepoCache

#
# Binary OPTIONS
#
#[bin]
#FileManager = vifm
#MFlags = --skippgpcheck
#Sudo = doas
```

**Comment by IPlayZed** [comment]
A bit of an update:

I tried creating and adding the line to the local gnupg configuration file, which paru uses:
_~/.gnupg/gpg.conf_:
```
keyserver hkps://keys.openpgp.org
```

The output is:
```shell
system@minefpc ~ [1]> paru -S linux-mainline
:: Resolving dependencies...
:: Calculating conflicts...
:: Calculating inner conflicts...

Aur (1) linux-mainline-6.7rc6-1

:: Proceed to review? [Y/n]: 

:: Downloading PKGBUILDs...
 PKGBUILDs up to date
 nothing new to review
:: keys need to be imported:
     ABAF11C65A2970B130ABE3C479BE3E4300411886 wanted by: linux-mainline-6.7rc6-1
:: import? [Y/n]: 
gpg: key 79BE3E4300411886: no user ID
gpg: Total number processed: 1
error: failed to run: gpg --recv-keys ABAF11C65A2970B130ABE3C479BE3E4300411886: 
```
[This issue](https://forum.endeavouros.com/t/solved-yay-cannot-import-key-from-keyserver-no-data-no-user-id/36311) seems to actually be the for the same key! However trying to specify the keyserver to the solution accepted there results in `gpg: keyserver receive failed: No keyserver available`.

It actually seems like this is not a paru problem, where should I open a bug for this?
I would personally not close this as some other people might see similar things, just in case.

**Comment by ac130kz** [comment]
Managed to solve it with an extra keyserver, there should be a few of them in the list, I guess.
```bash
gpg --keyserver hkps://pgp.surf.nl --recv-keys ABAF11C65A2970B130ABE3C479BE3E4300411886
```

**Comment by IPlayZed** [comment]
> Managed to solve it with an extra keyserver, there should be a few of them in the list, I guess.
> 
> ```shell
> gpg --keyserver hkps://pgp.surf.nl --recv-keys ABAF11C65A2970B130ABE3C479BE3E4300411886
> ```

So it really is hit or miss... This seems to have worked.

I think it would be cool if Paru was able to try some well-known working keyservers if the default one fails.

**Comment by IPlayZed** [comment]
This is closed, but I will implement this feature I mentioned. I think it would be a great improvement.

**Comment by stdolan** [comment]
@IPlayZed did this feature get implemented? Hit this same issue today on `paru v2.0.4 - libalpm v15.0.0`.

**Comment by IPlayZed** [comment]
> [@IPlayZed](https://github.com/IPlayZed) did this feature get implemented? Hit this same issue today on `paru v2.0.4 - libalpm v15.0.0`.

No such feature was ever implemented AFAIK. You can always add a new keyserver, like the comment suggests.

**Comment by stdolan** [comment]
> > [@IPlayZed](https://github.com/IPlayZed) did this feature get implemented? Hit this same issue today on `paru v2.0.4 - libalpm v15.0.0`.
> 
> No such feature was ever implemented AFAIK. You can always add a new keyserver, like the comment suggests.

Got it. Yep, that worked, I tweaked the config to use a different one.

Is there interest in that feature getting added? I didn't see an issue for it, but I like to try to give it a shot.

**Comment by IPlayZed** [comment]
@stdolan I am unaware of anyone who volunteered for it, and I have zero rust knowledge.
Based on the experiences we had so far, I would say we will not be the last ones who run into this, so if you could do it, go ahead :)
Is there a list of reliable key servers we could probe and add?
